### ISSUE TITLE
[training]loss customizing ideation

### ISSUE THREAD

**Issue by crystal0112** [sktaiflow/message-generator#2]
Bart 모델의 산출물인 hidden_state 혹은 logit의 값을 가지고 distance로 두 문장의 유사도를 파악
추가 Loss function에 관한 논의는 추가로 진행할 예정

**Comment by crystal0112** [comment]
참고할 만한 이슈
https://github.com/huggingface/transformers/issues/3001

Sentence 유사도 계산
console.cloud.google.com/bigquery?project=sktaic-datahub&j=bq:US:bquxjob_7a4fe11_17393f3b544&page=queryresults

**Comment by gowun** [comment]
일단 제가 추가한 건... 세가지 입니다
1. last_emc_emb_distance : token-pairwise encoder_last_hidden_states 간 평균 거리
2. logits_distance : token-pairwise logit 간 평균 거리
3. emc_emb_logits_distance: token-pairwise encoder_last_hidden_states 간 평균 거리 + token-pairwise logit 간 평균 거리
![image](https://user-images.githubusercontent.com/28796327/107170606-2b12d500-6a04-11eb-9eb2-d114b30999fd.png)
아직 수정님처럼 실제 모델을 통과시켰을 때 차이를 안 봐서.. 그거 추가로 여기 붙일게요.

**Comment by gowun** [comment]
수정님 similarity 노트북 봤구요. 우려되는 점 2가지입니다.
1. 위는 완전 다른 context, 아래는 비슷한 context/의미 임에도 0.1밖에 차이가 안나는 것..
![image](https://user-images.githubusercontent.com/28796327/107170690-601f2780-6a04-11eb-8cb9-ae3591a22111.png)
2. similarity의 range -1 ~ 1사이인데, 0에 가까울수록 context가 다르고, 절대값이 1에 가까울수록 context가 비슷, 그런데 -1 방향이라면 방향성이 다른 것.. 과연 이들을 어떻게 동일한 대소 관계에 두느냐인듯 해요. 
과연, 하늘을 나는 새가 0.345라고 -0.4371의 피부 최악보다 더 비슷한 문장이라고 볼것인가?? 
![image](https://user-images.githubusercontent.com/28796327/107170786-8c3aa880-6a04-11eb-9d00-f3831da39122.png)

**Comment by gowun** [comment]
오늘 회의 때 이야기 나온 nn.CosineEmbeddingLoss의 결과 고찰입니다. (y = 1로 둬서 -2 ~ 0의 값을 가짐)
https://github.com/sktaiflow/message-generator/blob/kennie/preprocessing/kennie/embedding_distance.ipynb

![image](https://user-images.githubusercontent.com/28796327/107188933-ddaa5e00-6a2b-11eb-8ea5-26df2c20381f.png)

위의 그림에서 3번째(encoder_last_hidden_states), 4번째(logits) 값만 참고하세요.

상식적으로??(인간이 이해하는 문맥상...) 3번, 5번 loss < 1번 loss < 2번, 6번 loss < 4번 loss 의 순으로 나와야 할 거 같은데 그렇지 않음..
- encoder_last_hidden_states : 3 < 1 < 5 < 2 < 4 < 6
- logits : 1 < 6 < 2 < 5 < 4 << 3 
(그나마, encoder_last_hidden_states가 순서가 비슷하고, logits은 완전히 다름!)

**Comment by gowun** [comment]
![image](https://user-images.githubusercontent.com/28796327/107302543-e1d48b00-6ac0-11eb-8c5b-10c06630dc25.png)
다행히 view를 통해 3d -> 2d로 바꿔도 cosine embedding loss 값은 동일함

**Comment by gowun** [comment]
심각한 문제 발견!! 
max_len 에 따른 값이 차이가 상당히 큼!! (특히, keywords 스타일은 max_len에 절대 근접하지 않으므로..)
아마도 필요한 만큼 데이터를 잘라서 cosine embedding loss를 계산해야 할듯 
![image](https://user-images.githubusercontent.com/28796327/107303993-85269f80-6ac3-11eb-8250-88e8766a2e4e.png)

**Comment by gowun** [comment]
@haven-jeon @soeque1 

contextually similarity 관련 논문 입니다. https://arxiv.org/pdf/1908.10084.pdf 

아래 표를 보시면, row는 sts 관련 모델 (첫번째 블록은 기존 bert모델 기반하여 각종 similarity 계산법을 사용, 두번째 세번째 블록은 sts fine tuning 한 모델들) 이고, column은 sts 용 public dataset 입니다.
저희는 이 중에, 현재 avg. bert embeddings 과 bert cls-vector 를 계속 논의해 왔고, 현재 https://github.com/sktaiflow/message-generator/blob/master/training/kennie/message_generator.py 내에 구현되어 있습니다.

향후, 이들을 가지고 실험할 계획이며 예상되는 2개의 문제점은
- cos_emb_loss의 range가 0 ~ 2 사이로 너무 작은 것 
- 저희 데이터 특성이 input은 아주 길수도 있고 (news 경우), output은 굉장히 짧은데(20 토큰 내) padding 된 부분이 어느정도 들어가냐에 따라 cos_emb_loss 값이 완전히 달라져서(padding이 길수록 0에 아주 가까워짐) 현재 코드 상은 label 길이에 맞춰 잘라 보는 걸로 해 놓았는데 (cross entropy와 비슷하게 처리), 만약에 label 길이는 짧은데 model(x)의 길이가 좀 더 길 경우 뒷부분이 잘리는 건 괜찮은지?? (이건 cross entropy 로도 마찬가지..)  아님 딱 정해놓고 앞에 5개 정도씩만 비교한다던지?? 
등의 고민들이 있습니다. 

![image](https://user-images.githubusercontent.com/28796327/107446211-4820e280-6b81-11eb-9c15-3326c3edb712.png)

**Comment by gowun** [comment]
현재까지 실험은 다음과 같이 수행함

1. nn.CosineEmbeddingLoss() between y.encoder_last_hidden_state and out.encoder_last_hidden_state
2. 1 - torch.mean(nn.CosineSimilarity(dim=1)) between y.encoder_last_hidden_state and out.encoder_last_hidden_state
3. log or exp of 1 - torch.mean(nn.CosineSimilarity(dim=1)) between y.encoder_last_hidden_state and out.encoder_last_hidden_state : to expand range up to 0 ~ more than 6
4. 1 - torch.mean(nn.CosineSimilarity(dim=1)) between y.logits and out.logits

- 4가지 모든 케이스에서 학습을 하면 할수록 어떤 문장간의 로스도 0으로 계산되기 때문에 targeted training이 아예 안된다.
- 애초에 아웃풋 그 자체가 아닌 내부(혹은 부분? 중간?) 결과 간 차이를 로스를 최소화하는 방향으로 학습시키므로 오히려 encoder network의 weights가 모든 문장에 대해 서로 구분이 안 가게끔 학습되는 것 같다.
- 결론적으로 학습 시 loss function으로는 사용가능해 보이지는 않으나, 2는 evaluation metrics으로써는 현재 생성된 문장이 기대되는 정답과 얼마나 유사한지 계산하는 것에는 유용하다.
